Adding CudaMallocSync support when using CUDA version >= 11.2

matt-stack opened the pull request:
https://github.com/kokkos/kokkos/issues/3981

Certain memory allocation patterns can take advantage of using CudaMallocAsync in CUDA 11.2 as opposed to using cudaMalloc. There is an immediate cudaDeviceSynchronize() call after the cudaMallocAsync in this implementation so that we can take advantage of the implicit pool allocators, and not introduce unintended asynchronous behavior. CudaFree has also been replaced by cudaFreeAsync when the ifdef confirms CUDA 11.2, so that the memory is returned to the pool, and a cudaDeviceSynchronize() right after.

dalg24-jenkins commented:
Can one of the admins verify this patch?

Rombur commented:
OK to test

crtrott commented:
Pushed the clang-format fix.

dalg24 commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -221,7 +221,14 @@ void *CudaSpace::impl_allocate(
   void *ptr = nullptr;
 
+#ifndef CUDART_VERSION
+#no CUDART_VERSION
```
```suggestion
#error CUDART_VERSION undefined!
```

dalg24 commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -339,7 +346,14 @@ void CudaSpace::impl_deallocate(
+#ifndef CUDART_VERSION
+#no CUDART_VERSION
+#elif (CUDART_VERSION >= 11020)
+    CUDA_SAFE_CALL(cudaFreeAsync(arg_alloc_ptr, 0));
+    cudaDeviceSynchronize();
```
I expect synchronization needs to happen before the deallocation.
```suggestion
    CUDA_SAFE_CALL(cudaDeviceSynchronize());
    CUDA_SAFE_CALL(cudaFreeAsync(arg_alloc_ptr, 0));
```

matt-stack replied:
Thanks for the comment, my thinking is that to replicate cudaFree using cudaFreeAsync and cudaDeviceSynchronize, the sync would go after the cudaFreeAsync call because if there is a use case that a user is rapidly allocating and deallocating, the sync would make sure that the allocation is in fact returned to the pool before the next allocation request. I don't believe there should be a sync before the cudaFreeAsync because there is a sync immediately right after the cudaDeviceSync, and as of right now we don't want to introduce any asynchronous behavior.

matt-stack replied:
Hi @dalg24, what are your thoughts on the comment above? I confirmed my suspicion with the CUDA team, and they actually mentioned that the sync was more or less functionally equivalent to before (or not at all, as long as the rest of the code does not have any loose async calls) as the memory is not freed until the prior GPU work completes, but there is a small perf bonus of including it after the call to free.

maxpkatz replied:
@crtrott told me that memory allocations in Views are semantically fencing in Kokkos (at least as of now), so I think that for safety we should synchronize both before and after the free so that it's clear to the reader that no actually asynchronous workflows involving memory allocation are currently possible.

matt-stack replied:
Good call, I think that is a good add for readability, added and committed in https://github.com/kokkos/kokkos/pull/4026/commits/ac17a44cd61fabe88915db998118f8782e81e347

crtrott replied:
so cudaFree and cudaMalloc do block all outstanding kernels right? Do cudaFreeAsync and cudaMallocAsync do so too? If not yes, you need to have that device syncrhonization before.

dalg24 commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -339,7 +346,14 @@ void CudaSpace::impl_deallocate(
 
   try {
+#ifndef CUDART_VERSION
+#no CUDART_VERSION
```
Same comment as above.  `#no` is not a preprocessor directive as far as I know.

matt-stack replied:
Thanks for the note, changed here: https://github.com/kokkos/kokkos/pull/4026/commits/3d80e7166e40d30d75d358a9565d18f6bb8dbe4f

masterleinad commented:
Retest this please.

masterleinad commented:
Retest this please.

matt-stack commented:
@masterleinad Testing now

masterleinad commented:
Retest this please.

matt-stack commented:
Update: I am working on one small change regarding a small "bug" (that has been fixed in the most recent CUDA). Hoping to push for review and thoughts today

matt-stack commented:
A Cuda bug came up when I was running the unit tests, and it requires a small addition to CudaSpace to avoid. I am adding it here to this PR for thoughts. The bug is that if you request a number very close to the numerical limit for size_t, internally this number is rounded up to SIZE_MAX+1, which gets used by some code that does not have a check for 0, and segfaults. It was reported that this fix will be in Cuda 11.4.

This came up from KokkosCore_UnitTest_Cuda2, where one test allocates 
`auto arg_alloc_size = std::numeric_limits<size_t>::max() - 42;`

The check does reduce the readability of this section of CudaSpace with an extra if-else statement, and after much thought I am curious if it should be included. The case where the bug is triggered seems unlikely (requesting close to SIZE_MAX of size_t, or a small negative size_t which is the same case). I can revert it back to the original version if these additions are unnecessary. @crtrott @maxpkatz

masterleinad commented:
The indentation needs to be fixed:
```
./scripts/docker/check_format_cpp.sh
diff --git a/core/src/Cuda/Kokkos_CudaSpace.cpp b/core/src/Cuda/Kokkos_CudaSpace.cpp
index 6123e93cf..9b8aaef68 100644
--- a/core/src/Cuda/Kokkos_CudaSpace.cpp
+++ b/core/src/Cuda/Kokkos_CudaSpace.cpp
@@ -226,11 +226,10 @@ void *CudaSpace::impl_allocate(
 #error CUDART_VERSION undefined!
 #elif (CUDART_VERSION >= 11020)
   cudaError_t error_code;
-  if ( (size_t)arg_alloc_size < (std::numeric_limits<size_t>::max() - 1000) ) {
+  if ((size_t)arg_alloc_size < (std::numeric_limits<size_t>::max() - 1000)) {
     error_code = cudaMallocAsync(&ptr, arg_alloc_size, 0);
     cudaDeviceSynchronize();
-  }
-  else {
+  } else {
     error_code = cudaErrorInvalidValue;
   }
 #else
script returned exit code 1
```

masterleinad commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -221,7 +222,20 @@ void *CudaSpace::impl_allocate(
+  if ( (size_t)arg_alloc_size < (std::numeric_limits<size_t>::max() - 1000) ) {
+    error_code = cudaMallocAsync(&ptr, arg_alloc_size, 0);
+    cudaDeviceSynchronize();
+  }
+  else {
+    error_code = cudaErrorInvalidValue;
+  }
+#else
```
Since this is a performance improvement, I would fall back to the original approach, i.e. using `cudaMalloc`, if the bug is triggered. Also, we should add a FIXME with an explanation so that we can remove the restriction eventually.

matt-stack replied:
I agree, I will make these changes now

matt-stack replied:
@masterleinad Thanks for your comments, I added these changes

masterleinad commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -221,7 +222,20 @@ void *CudaSpace::impl_allocate(
+  // FIXME_CUDA Checks for bug (info in PR 4026, fixed in Cuda 11.4)
+  if ((size_t)arg_alloc_size < (std::numeric_limits<size_t>::max() - 1000)) {
+    error_code = cudaMallocAsync(&ptr, arg_alloc_size, 0);
+    cudaDeviceSynchronize();
```
```suggestion
    CUDA_SAFE_CALL(cudaDeviceSynchronize());
```

masterleinad commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -339,7 +353,15 @@ void CudaSpace::impl_deallocate(
+#ifndef CUDART_VERSION
+#error CUDART_VERSION undefined!
+#elif (CUDART_VERSION >= 11020)
+    cudaDeviceSynchronize();
```
```suggestion
    CUDA_SAFE_CALL(cudaDeviceSynchronize());
```

masterleinad commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -339,7 +353,15 @@ void CudaSpace::impl_deallocate(
+#elif (CUDART_VERSION >= 11020)
+    cudaDeviceSynchronize();
+    CUDA_SAFE_CALL(cudaFreeAsync(arg_alloc_ptr, 0));
+    cudaDeviceSynchronize();
```
```suggestion
    CUDA_SAFE_CALL(cudaDeviceSynchronize());
```

masterleinad reviewed (commented):
Please also wrap `cudaDeviceSynchronize` as suggested below.

crtrott commented:
Retest this please

matt-stack commented:
Are there any more tests that should be run/change requests?

matt-stack commented:
Hi @dalg24, would you be able to check out this current state for approval? Thanks!

dalg24 commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -221,7 +222,20 @@ void *CudaSpace::impl_allocate(
+#error CUDART_VERSION undefined!
+#elif (CUDART_VERSION >= 11020)
+  cudaError_t error_code;
+  // FIXME_CUDA Checks for bug (info in PR 4026, fixed in Cuda 11.4)
```
Or is it a bug introduced in 11.2 and fixed in 11.4?

crtrott reviewed (changes requested):
Hi guys,

I did some performance checking and for small allocations this seems to be slower.

I ran this code, which does a bunch of random sizes allocations, deletes them, and does that in a loop:

```c++
#include <Kokkos_Core.hpp>
#include <cmath>
#include <cstdlib>

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc, argv);
  {
    int N = argc > 1 ? atoi(argv[1]) : 1000;
    int R = argc > 2 ? atoi(argv[2]) : 10;
    int MAX_SIZE = argc > 3 ? atoi(argv[3]) : 10000000;

    double** ptrs = new double*[R];
    srand(5123);
    Kokkos::Timer timer;
    for(int i=0;i<N;i++) {
      for(int r=0; r<R; r++) {
        int size = rand()%MAX_SIZE;
        ptrs[r]=(double*)Kokkos::kokkos_malloc<>(size*8);
      }
      for(int r=0; r<R; r++) Kokkos::kokkos_free<>(ptrs[r]);
    }
    printf("%lf\n",R*N/timer.seconds());
  }
  Kokkos::finalize();
}
```

At the end it spits out an allocation/deallocation rate.
With develop for N=1000, R=10 and MAX_SIZE=100 i got around 70k/s while with this PR I get only 54k/s.
With MAX_SIZE=10,000,000 I get 875/s and 899/s respectively (i.e. the new code is faster).

With MAX_SIZE=10,000 I still get 67k vs 53k.
With MAX_SIZE=100,000 however its 18k vs 52k
With MAX_SIZE=1,000,000 its 4k vs 9k

Note Size is size in number of doubles. So it looks like <100kB cudaMalloc might be faster, while above that the async thing is faster. We probably should have a switchover in the code, the current one is also a bit weird? What is the business with std::numeric_limits<size_t>::max there?? even -1000 that limit is unrealstic, not even Summit has that much memory on the entire machine, not to speak of a single node.

matt-stack commented:
@crtrott I ran your test case and saw similar results on V100 with CUDA 11.2, I think a switch of cudaMalloc for <100kb and cudaMallocAsync for greater is a great idea. I can add that in for review. Yeah so for the std::numeric_limits<size_t>::max if statement, that is there because there is a bug in cudaMallocAsync that if you request a number close to SIZE_MAX then it will throw a segfault. It came up after running one of the Kokkos Cuda unit tests (the specific test request: auto alloc_size = std::numeric_limits<size_t>::max() - 42;) so I thought I would include the size check as it was a case in the units. The size statement check can easily be taken out, I agree that a user would already have to be in dangerous territory to trigger it (requesting unrealistic size, or a negative int, etc)

matt-stack commented:
@crtrott Hi Christian, I added a memory threshold of a requested alloc size based on testing I ran. I found that having any request less than 5 kb use cudaMalloc and larger use cudaMallocAsync was the best mix for the test case. I also removed the "bug check" if statement (detailed above), but could easily put it back in.

crtrott commented on core/src/Cuda/Kokkos_CudaSpace.cpp:
```diff
@@ -204,6 +204,8 @@ CudaUVMSpace::CudaUVMSpace() : m_device(Kokkos::Cuda().cuda_device()) {}
 
 CudaHostPinnedSpace::CudaHostPinnedSpace() {}
 
+int memory_threshold_g = 40000; // 5 kB
```
was this supposed to be 5kB or 40? otherwise its fine.

matt-stack replied:
Typo, the comment was supposed to be 40
